src/kakarot/instructions/stop_and_math_operations.cairo

```diff
@@ -240,15 +240,40 @@ namespace StopAndMathOperations {
         jmp end;
 
         addmod_not_zero:
-        // (a + b) mod n  = (a mod n + b mod n)mod n
-        let (_, a_mod_n) = uint256_unsigned_div_rem(popped[0], popped[2]);
-        let (_, b_mod_n) = uint256_unsigned_div_rem(popped[1], popped[2]);
-        let (sum, carry) = uint256_add(a_mod_n, b_mod_n);
-        let (_, result) = uint256_unsigned_div_rem(sum, popped[2]);
+        // (a + b) mod n  = (a mod n + b mod n) mod n
+        let (_, x) = uint256_unsigned_div_rem(popped[0], popped[2]);
+        let (_, y) = uint256_unsigned_div_rem(popped[1], popped[2]);
+        // x, y in range [0, n-1] thus:
+        // if x + y < n then x + y mod n = x + y
+        // if x + y >= n then x + y mod n = x + y - n
+        let (sum, carry) = uint256_add(x, y);
+
+        if (carry != 0) {
+            // result = (2**256) - (n - overflown_sum)
+            // <=> result = (2**256 - 1) - (n - overflown_sum - 1)
+            // as n > overflown_sum we can't have an underflow
```

Comment:
this is not so easy to follow.

You could

```
let a = popped[0]
let b = popped[1]
let n = popped[2]
```

then using your notations (the names of the variable)

```
result = 2**256 * carry + sum = 2**256 + sum
 = 2**256 + sum - n = 2**256 - 1 - n + sum + 1
```

and this is because `carry > 0 => n > 2**128` so `2**256 mod n = 2**256 - n`